### PR TITLE
Impl bridge token cap for security reason

### PR DESCRIPTION
**Pull request by rjman-ljm**
CHANGELOG:
> Scope: Bridge
1. Add `CapOrigin` to do high frequency operation(clean cap)
2. Add three calls
  - `set_bridge_token_status()`
  - `set_bridge_token_cap()`
  - `clean_cap_accumulated_value()`
3. Enrich `BridgeToken` structure and add `BridgeType`
  - `in_cap, out_cap`
  - `in_amount, out_amount`
4. add tests
5. fix benchmarking
6. update launch script

### CODE REVIEW

File: pallets/bridge/src/lib.rs [hunk 1]
```diff
@@ -185,6 +201,18 @@ pub mod pallet {
         /// [bridge_token_id, fee]
         BridgeTokenFeeChanged(CurrencyId, BalanceOf<T>),
 
+        /// The status of the bridge token has changed
+        /// [bridge_token_id, enabled]
+        BridgeTokenStatusChanged(CurrencyId, bool),
```
Comment (GopherJ) [inline pallets/bridge/src/lib.rs:206]: Just wondering if we should use the same word in different place

Changed vs Updated

File: pallets/bridge/src/lib.rs [hunk 2]
```diff
@@ -399,6 +428,80 @@ pub mod pallet {
...
+        /// Clean the accumulated cap value to make bridge work again
+        #[pallet::weight(T::WeightInfo::clean_cap_accumulated_value())]
+        #[transactional]
+        pub fn clean_cap_accumulated_value(
+            origin: OriginFor<T>,
+            bridge_token_id: CurrencyId,
+            bridge_type: BridgeType,
+        ) -> DispatchResult {
+            T::CapCleanOrigin::ensure_origin(origin)?;
+            Self::ensure_bridge_token_registered(bridge_token_id)?;
```
Comment (GopherJ) [inline pallets/bridge/src/lib.rs:488]: this is duplicated multiple time, we can have a common function which accepts a modifier fn

File: pallets/bridge/src/lib.rs [hunk 3]
```diff
@@ -423,18 +526,26 @@ pub mod pallet {
             Self::ensure_bridge_token_registered(bridge_token_id)?;
             Self::ensure_amount_valid(amount)?;
 
-            let asset_id = AssetIds::<T>::get(bridge_token_id);
-            let BridgeToken { external, fee, .. } = BridgeTokens::<T>::get(asset_id);
-            let actual_amount = amount
-                .checked_sub(fee)
-                .ok_or(Error::<T>::BridgedAmountTooLow)?;
+            let asset_id = Self::asset_ids(bridge_token_id);
```
Comment (GopherJ) [inline pallets/bridge/src/lib.rs:529]: the getter can be more clear `Self::asset_id`

File: pallets/bridge/src/lib.rs [hunk 4]
```diff
@@ -423,18 +526,26 @@ pub mod pallet {
...
+            let BridgeToken {
+                external,
+                fee,
+                enable,
+                ..
+            } = Self::bridge_tokens(asset_id);
```
Comment (GopherJ) [inline pallets/bridge/src/lib.rs:535]: same as above, `Self::bridge_token`

File: pallets/bridge/src/lib.rs [hunk 5]
```diff
@@ -621,6 +737,41 @@ impl<T: Config> Pallet<T> {
...
+        amount: BalanceOf<T>,
+        bridge_type: BridgeType,
+    ) -> Result<Balance, Error<T>> {
+        let bridge_token = Self::bridge_tokens(asset_id);
+        let new_amount = match bridge_type {
+            BridgeType::BridgeOut => {
+                let new_out_amount = bridge_token
+                    .out_amount
+                    .checked_add(amount)
+                    .ok_or(Error::<T>::BridgeAmountOverflow)?;
```
Comment (GopherJ) [inline pallets/bridge/src/lib.rs:752]: looks like `ArithmeticError` is enough to replace this one?

File: pallets/bridge/src/lib.rs [hunk 6]
```diff
@@ -621,6 +737,41 @@ impl<T: Config> Pallet<T> {
...
+                ensure!(
+                    new_out_amount <= bridge_token.out_cap,
+                    Error::<T>::BridgeOutCapacityExceeded
```
Comment (GopherJ) [inline pallets/bridge/src/lib.rs:755]: BridgeOutCapExceeded

File: pallets/bridge/src/lib.rs [hunk 7]
```diff
@@ -621,6 +737,41 @@ impl<T: Config> Pallet<T> {
...
+                new_out_amount
+            }
+            BridgeType::BridgeIn => {
+                let new_in_amount = bridge_token
+                    .in_amount
+                    .checked_add(amount)
+                    .ok_or(Error::<T>::BridgeAmountOverflow)?;
+                ensure!(
+                    new_in_amount <= bridge_token.in_cap,
+                    Error::<T>::BridgeInCapacityExceeded
```
Comment (GopherJ) [inline pallets/bridge/src/lib.rs:766]: BridgeInCapExceeded

File: pallets/bridge/src/lib.rs [hunk 8]
```diff
@@ -84,6 +85,9 @@ pub mod pallet {
         /// Root can always do this.
         type OperateOrigin: EnsureOrigin<Self::Origin>;
 
+        /// The origin which can clean accumulated cap value
+        type CapCleanOrigin: EnsureOrigin<Self::Origin>;
```
Comment (GopherJ) [inline pallets/bridge/src/lib.rs:89]: CapOrigin to make it more general

File: pallets/bridge/src/lib.rs [hunk 9]
```diff
@@ -84,6 +85,9 @@ pub mod pallet {
         /// Root can always do this.
         type OperateOrigin: EnsureOrigin<Self::Origin>;
```
Comment (GopherJ) [inline pallets/bridge/src/lib.rs:86]: OperateOrigin looks not clear

File: pallets/bridge/src/lib.rs [hunk 10]
```diff
@@ -423,18 +526,26 @@ pub mod pallet {
...
             if external {
                 T::Assets::burn_from(asset_id, &who, amount)?;
                 T::Assets::mint_into(asset_id, &Self::account_id(), fee)?;
             } else {
                 T::Assets::transfer(asset_id, &who, &Self::account_id(), amount, false)?;
             }
 
+            let actual_amount = amount
+                .checked_sub(fee)
+                .ok_or(Error::<T>::BridgedAmountTooLow)?;
```
Comment (GopherJ) [inline pallets/bridge/src/lib.rs:548]: ArithmeticError is enough?

File: pallets/bridge/src/lib.rs (hunk 1 above)
Comment (rjman-ljm) [reply pallets/bridge/src/lib.rs:206]: yep

File: pallets/bridge/src/lib.rs (hunk 2 above)
Comment (rjman-ljm) [reply pallets/bridge/src/lib.rs:488]: good idea

File: pallets/bridge/src/lib.rs (hunk 9 above)
Comment (rjman-ljm) [reply pallets/bridge/src/lib.rs]: I split it into `UpdateChainOrigin` and `UpdateTokenOrigin`

**Comment by GopherJ** [comment]
@zhoujia6139 can you also give a review

**Pull request merged**